src/components/LiquidityDepth/DensityChart.tsx

```diff
@@ -0,0 +1,204 @@
+import { useDensityChartData, ChartEntry, ChartContext } from './hooks'
+import { Bound } from 'state/mint/v3/actions'
+
+const sampleData: Partial<ChartEntry>[] = [
+  { price0: 0, activeLiquidity: 1 },
+  { price0: 1, activeLiquidity: 2 },
+  { price0: 2, activeLiquidity: 3 },
+  { price0: 3, activeLiquidity: 6 },
+  { price0: 4, activeLiquidity: 5 },
+  { price0: 5, activeLiquidity: 3 },
+  { price0: 6, activeLiquidity: 2 },
+]
+
+const Wrapper = styled(Box)`
+  position: relative;
+  height: 250px;
+
+  display: grid;
+  justify-content: center;
+  align-content: center;
+`
+
+export default function DensityChart({
+  currencyA,
+  currencyB,
+  feeAmount,
+  price,
+  priceLower,
+  priceUpper,
+  onLeftRangeInput,
+  onRightRangeInput,
+  interactive,
+  ticksAtLimit,
+}: {
+  currencyA: Currency | undefined
+  currencyB: Currency | undefined
+  feeAmount?: number
+  price?: Price<Token, Token>
+  priceLower?: Price<Token, Token>
+  priceUpper?: Price<Token, Token>
+  onLeftRangeInput: (typedValue: string) => void
+  onRightRangeInput: (typedValue: string) => void
+  interactive: boolean
+  ticksAtLimit: { [bound in Bound]?: boolean | undefined }
+}) {
+  const { zoom } = useContext(ChartContext)
+
+  const theme = useTheme()
+
+  const tokenAColor = useColor(currencyA?.wrapped)
+  const tokenBColor = useColor(currencyB?.wrapped)
+
+  const { loading, activeChartEntry, maxLiquidity, formattedData } = useDensityChartData({
+    currencyA,
+    currencyB,
+    feeAmount,
+  })
+
+  const isSorted = currencyA && currencyB && currencyA?.wrapped.sortsBefore(currencyB?.wrapped)
+
+  const leftPrice = isSorted ? priceLower : priceUpper?.invert()
+  const rightPrice = isSorted ? priceUpper : priceLower?.invert()
+
+  if (loading) {
+    return (
+      <Wrapper>
+        <Loader stroke={theme.text4} />
+      </Wrapper>
+    )
+  }
+
+  interactive = interactive && Boolean(formattedData?.length)
+
+  const filteredData =
+    activeChartEntry && zoom
+      ? formattedData?.slice(Math.max(0, activeChartEntry.index - zoom), activeChartEntry.index + zoom)
+      : undefined
+
+  return (
+    <Wrapper>
+      {/* filteredData === undefined will show sample data */}
+      {filteredData === [] ? (
+        <ColumnCenter>
+          <XCircle stroke={theme.text4} />
+          <TYPE.darkGray padding={10}>
+            <Trans>No data</Trans>
+          </TYPE.darkGray>
+        </ColumnCenter>
+      ) : (
+        <VictoryChart
+          height={275}
+          padding={40}
+          minDomain={{ y: 0 }}
+          containerComponent={
+            <VictoryBrushContainer
+              allowDraw={false}
+              allowDrag={interactive}
+              allowResize={interactive}
+              brushDimension="x"
+              brushDomain={
+                leftPrice && rightPrice
+                  ? {
+                      x: [parseFloat(leftPrice?.toSignificant(5)), parseFloat(rightPrice?.toSignificant(5))],
```

Comment:
Will this cause issues with really weird pairs? like a BTC/X pair with >5 decimals?

Reply:
What's your suggestion here? Should I go to 8?